Flutter_TDD_Architecture_Course_notes.md

```diff
@@ -277,3 +277,740 @@ class GetRandomNumberTrivia extends UseCase<NumberTrivia, NoParams> {
+
+class MockLocalDataSource extends Mock implements NumberTriviaLocalDataSource {}
+
+class MockNetworkInfo extends Mock implements NetworkInfo {}
+
+void main() {
+  NumberTriviaRepositoryImpl repository;
+  MockRemoteDataSource mockRemoteDataSource;
+  MockLocalDataSource mockLocalDataSource;
+  MockNetworkInfo mockNetworkInfo;
+
+  setUp(() {
+    mockRemoteDataSource = MockRemoteDataSource();
+    mockLocalDataSource = MockLocalDataSource();
+    mockNetworkInfo = MockNetworkInfo();
+    repository = NumberTriviaRepositoryImpl(
+      remoteDataSource: mockRemoteDataSource,
+      localDataSource: mockLocalDataSource,
+      networkInfo: mockNetworkInfo,
+    );
+  });
+}
+```
+If we run the test now it will give an error because have not yet written the implementation. 
+Open the `number_trivia_repository_impl.dart` file and add the following code:
+
+```dart
+import 'package:dartz/dartz.dart';
+import 'package:meta/meta.dart';
+
+import '../../../../core/error/failure.dart';
+import '../../../../core/platform/network_info.dart';
+import '../../domain/entities/number_trivia.dart';
+import '../../domain/repositories/number_trivia_repository.dart';
+import '../datasources/number_trivia_local_data_source.dart';
+import '../datasources/number_trivia_remote_data_source.dart';
+
+class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
+  final NumberTriviaRemoteDataSource remoteDataSource;
+  final NumberTriviaLocalDataSource localDataSource;
+  final NetworkInfo networkInfo;
+
+  NumberTriviaRepositoryImpl({
+    @required this.remoteDataSource,
+    @required this.localDataSource,
+    @required this.networkInfo,
+  });
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
+    // TODO: implement getConcreteNumberTrivia
+    return null;
+  }
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
+    // TODO: implement getRandomNumberTrivia
+    return null;
+  }
+}
+```
+#### 6. Repository Implementation
+> Video 6: https://www.youtube.com/watch?v=bfEKPKKy9dA
+
+In the `sixth` tutorial start by explaining that we will have to know if the user has an internet connection and for that a test will be done.
+
+#### Device Is Online
+
+```dart
+group('getConcreteNumberTrivia', () {
+  // DATA FOR THE MOCKS AND ASSERTIONS
+  // We'll use these three variables throughout all the tests
+  final tNumber = 1;
+  final tNumberTriviaModel =
+      NumberTriviaModel(number: tNumber, text: 'test trivia');
+  final NumberTrivia tNumberTrivia = tNumberTriviaModel;
+
+  test('should check if the device is online', () {
+    //arrange
+    when(mockNetworkInfo.isConnected).thenAnswer((_) async => true);
+    // act
+    repository.getConcreteNumberTrivia(tNumber);
+    // assert
+    verify(mockNetworkInfo.isConnected);
+  });
+});
+```
+The test will `fail` because we have to add features.
+Add the code to the `number_trivia_repository_impl_test.`dart.
```

Comment:
Miguel, you are _clearly_ not reading what you write ***`before`*** assigning it to me.
If you did, you would spot and _fix_ these kinds of _basic_ errors:
```suggestion
Add the code to the `number_trivia_repository_impl_test.dart` file:
```